packages/framework/src/client.ts

```diff
@@ -309,11 +313,20 @@ export class Client {
 
       const step = this.getStep(event.workflowId, stepId);
       const isPreview = event.action === PostActionEnum.PREVIEW;
+      let controls = {};
+
+      if (stepId === event.stepId) {
+        const templateControls = await this.createStepControls(step, event);
+
+        if (this.compileControls) {
+          controls = await this.renderTemplateControls(templateControls, event);
+        } else {
+          controls = templateControls;
+        }
+      }
 
       // Only evaluate a skip condition when the step is the current step and not in preview mode.
       if (!isPreview && stepId === event.stepId) {
-        const templateControls = await this.createStepControls(step, event);
-        const controls = await this.compileControls(templateControls, event);
```

Comment:
This change removes the second (for all client users) or third (for email Novu users) duplicate control compilation, simplifying the process.